tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/ral_legalize_to_llvm.cc

```diff
@@ -0,0 +1,556 @@
+  } else if (auto ctx_type = t.dyn_cast<RalExecutionContextType>() ||
+                             t == llvm_pointer_type) {
+    out.append("pvoid");
+  } else if (t == llvm_pointer_pointer_type) {
+    out.append("ppvoid");
+  } else if (t.isIndex()) {
+    // index is mapping to int64_t a.t.m. Re-visit this in case necessary.
+    out.append("i64");
+  } else {
+    // unknown type
+    return failure();
+  }
+  return success();
+}
+
+// Encodes a ral_dispatch op and appends the encoding to the string buffer
+// `out`. The format:
+//   encoding = separator.join(target_name, device, inputs_encode,
+//   outputs_encode)
+//
+//   separator = '___'
+//
+//   target_name: name of the external function to dispatch.
+//
+//   device: user defined string (e.g. cpu or gpu)
+//
+//   inputs_encode = type_separator.join([type_encoding for type in
+//   input_types])
+//
+//   outputs_encode = type_separator.join([type_encoding for type
+//   in output_types])
+//
+//   type_separator = '_'
+LogicalResult getDispatchOpSignatureEncoding(DispatchOp dispatch_op,
+                                             StrT& out) {
+  const char* separator = "___";
+  // append signature prefix
+  out.append(dispatch_op.call_target_name());
+  out.append(separator);
+
+  // encode backend (device) info
+  out.append(dispatch_op.backend_config());
+  out.append(separator);
+
+  // encode input types
+  Operation* op = dispatch_op.getOperation();
+  for (auto& en : llvm::enumerate(op->getOperandTypes())) {
+    if (en.index() != 0) out.append("_");
+    if (failed(getTypeEncoding(op->getContext(), en.value(), out)))
+      return failure();
+  }
+  out.append(separator);
+
+  // encode output types
+  for (auto& en : llvm::enumerate(op->getResultTypes())) {
+    if (en.index() != 0) out.append("_");
+    if (failed(getTypeEncoding(op->getContext(), en.value(), out)))
+      return failure();
+  }
+  if (!op->getNumResults()) out.append("void");
+  return success();
+}
+
+// Loads a global op at the current insertion point and returns the loaded
+// value.
+Value loadGlobalString(OpBuilder& builder, const Location& loc,
+                       GlobalOp globalOp) {
+  MLIRContext* ctx = builder.getContext();
+  Value globalPtr = builder.create<LLVM::AddressOfOp>(loc, globalOp);
+  Value cst0 = builder.create<LLVM::ConstantOp>(
+      loc, IntegerType::get(ctx, 64),
+      builder.getIntegerAttr(builder.getIndexType(), 0));
+  return builder.create<LLVM::GEPOp>(
+      loc, LLVM::LLVMPointerType::get(IntegerType::get(ctx, 8)), globalPtr,
+      ValueRange{cst0, cst0});
+}
+
+// Creates a global const string op named `name` using the value if not exists
+// and returns the Loaded value of this global op.
+Value loadOrCreateGlobalString(PatternRewriter& rewriter, Operation* op,
+                               StringRef name, StringRef value) {
+  ModuleOp module = op->getParentOfType<ModuleOp>();
+  GlobalOp globalOp = module.lookupSymbol<GlobalOp>(name);
```

Comment:
These lookups will be expensive, can you instead build a `SymbolTable` at the beginning of the function execution and update it through?
(Patterns would have to take it in their constructor and keep a reference to it)

Reply:
Thanks! I'm not sure if I fully understand what you mean.  I guess that you say `module.lookupSymbol` is expensive because this method needs to create new a `SymbolTable` every time it's called. 

It's worth noting that `RalToLLVMPass` is a function-level instead module-level pass to enable maximum parallelism, thus multiple instance of RalToLLVMPass may run simultaneously.  Therefore, we have to re-create a SymbolTable every time we need to lookup a symbol since other threads may insert symbols as well.

Reply:
Uh that seems like a race condition: actually function-level pass can't create symbol in the outer module like that.

Reply:
Now that there is no threading involved, can we look into this?

Reply:
I was waiting  for you here?

Reply:
Sorry... I misses the updates here. I have fixed it.